components/script/dom/readablebytestreamcontroller.rs

```diff
@@ -54,3 +176,327 @@ impl ReadableByteStreamControllerMethods for ReadableByteStreamController {
+pub fn setup_readable_byte_stream_controller_from_underlying_source(
+    cx: SafeJSContext,
+    stream: DomRoot<ReadableStream>,
+    underlying_source_obj: SafeHandleObject,
+    underlying_source_dict: UnderlyingSource,
+    highwatermark: f64,
+) -> Fallible<()> {
+    // Step 8
+    let auto_allocate_chunk_size = underlying_source_dict.autoAllocateChunkSize;
+    // Step 9
+    if let Some(size) = auto_allocate_chunk_size {
+        if size == 0 {
+            return Err(Error::Type("autoAllocateChunkSize can't be 0".to_string()));
+        }
+    }
+
+    // Step 2. - 7. See UnderlyingSourceAlgorithms
+    let algorithms = UnderlyingSourceAlgorithms::new(underlying_source_dict, underlying_source_obj);
+
+    // Step 1
+    let controller = ReadableByteStreamController::new(&*stream.global(), stream);
+
+    set_up_readable_byte_stream_controller(
+        cx,
+        controller,
+        ControllerAlgorithms::UnderlyingSource(algorithms),
+        highwatermark,
+        auto_allocate_chunk_size,
+    )
+}
+
+/// <https://streams.spec.whatwg.org/#set-up-readable-byte-stream-controller>
+fn set_up_readable_byte_stream_controller(
+    cx: SafeJSContext,
+    controller: DomRoot<ReadableByteStreamController>,
+    algorithms: ControllerAlgorithms,
+    highwatermark: f64,
+    auto_allocate_chunk_size: Option<u64>,
+) -> Fallible<()> {
+    // Step 1
+    assert!(controller.stream.controller().is_none());
+    // Step 2
+    if let Some(size) = auto_allocate_chunk_size {
+        assert!(size > 0);
+    }
+    // Step 3 is done in ReadableStreamDefaultController::new already.
+    // Step 4
+    controller.pull_again.set(false);
+    controller.pulling.set(false);
+    // Step 5 is done in ReadableStreamDefaultController::new already.
+    // Step 6
+    controller.queue.borrow_mut().clear();
+    // Step 7
+    controller.started.set(false);
+    controller.close_requested.set(false);
+    // Step 8
+    controller.strategy_highwatermark.set(highwatermark);
+    // Step 9 & 10
+    *controller.algorithms.borrow_mut() = algorithms;
+    // Step 11
+    controller
+        .auto_allocate_chunk_size
+        .set(auto_allocate_chunk_size);
+    // Step 12
+    controller.pending_pull_intos.borrow_mut().clear();
+    // Step 13
+    controller
+        .stream
+        .set_controller(ReadableStreamController::ReadableByteStreamController(
+            controller.clone(),
+        ));
+    // Step 14
+    rooted!(in(*cx) let mut start_result = UndefinedValue());
+    controller.algorithms.borrow().start(
+        cx,
+        ReadableStreamController::ReadableByteStreamController(controller.clone()),
+        start_result.handle_mut(),
+    )?;
+    // Step 15
+    let global = &*controller.stream.global();
+    let realm = enter_realm(&*global);
+    let comp = InRealm::Entered(&realm);
+    let start_promise = Promise::new_resolved(global, cx, start_result.handle())?;
+    // Step 16 & 17
+    start_promise.append_native_handler(
```

Comment:
Resolving a promise in the controller seems a bit complicated. Is it correct to append the handler after `new_resolved`?